nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp

```diff
@@ -0,0 +1,573 @@
+    if (unbounded_vel < min_approach_linear_velocity_) {
+      linear_vel = min_approach_linear_velocity_;
+    } else {
+      linear_vel = linear_vel * velocity_scaling;
+    }
+  }
+
+  // Limit linear velocities to be kinematically feasible, v = v0 + a * dt
+  double & dt = control_duration_;
+  const double max_feasible_linear_speed = curr_speed.linear.x + max_linear_accel_ * dt;
+  const double min_feasible_linear_speed = curr_speed.linear.x - max_linear_decel_ * dt;
+  linear_vel = std::clamp(linear_vel, min_feasible_linear_speed, max_feasible_linear_speed);
+
+  // Note(stevemacenski): for the moment, no smoothing on angular velocities, we find the existing
+  // commands are very smooth and we don't want to artifically reduce them if the hardware
+  // can handle it. Plus deviating from the commands here can be collision-inducing if users
+  // don't properly set these values, so hiding that complexity from them that would likely
+  // become the #1 cause of issues.
+  // const double max_feasible_angular_speed = curr_speed.angular.z + angular_accel_ * dt;
+  // const double min_feasible_angular_speed = curr_speed.angular.z - angular_accel_ * dt;
+  // angular_vel = std::clamp(angular_vel, min_feasible_angular_speed, max_feasible_angular_speed);
+
+  // Limit range to generally valid velocities
+  linear_vel = std::clamp(linear_vel, 0.0, desired_linear_vel_);
+}
+
+void RegulatedPurePursuitController::setPlan(const nav_msgs::msg::Path & path)
+{
+  global_plan_ = path;
+}
+
+nav_msgs::msg::Path RegulatedPurePursuitController::transformGlobalPlan(
+  const geometry_msgs::msg::PoseStamped & pose)
+{
+  if (global_plan_.poses.empty()) {
+    throw nav2_core::PlannerException("Received plan with zero length");
+  }
+
+  // let's get the pose of the robot in the frame of the plan
+  geometry_msgs::msg::PoseStamped robot_pose;
+  if (!transformPose(global_plan_.header.frame_id, pose, robot_pose))
+  {
+    throw nav2_core::PlannerException("Unable to transform robot pose into global plan's frame");
+  }
+
+  // We'll discard points on the plan that are outside the local costmap
+  nav2_costmap_2d::Costmap2D * costmap = costmap_ros_->getCostmap();
+  const double max_costmap_dim = std::max(costmap->getSizeInCellsX(), costmap->getSizeInCellsY());
+  const double max_transform_dist =  max_costmap_dim * costmap->getResolution() / 2.0;
+
+  // First find the closest pose on the path to the robot
+  auto transformation_begin =
+    min_by(
+    global_plan_.poses.begin(), global_plan_.poses.end(),
+    [&robot_pose](const geometry_msgs::msg::PoseStamped & ps) {
+      return euclidean_distance(robot_pose, ps);
+    });
+
+  // Find points definitely outside of the costmap so we won't transform them.
+  auto transformation_end = std::find_if(
+    transformation_begin, end(global_plan_.poses),
+    [&](const auto & global_plan_pose) {
+      return euclidean_distance(robot_pose, global_plan_pose) > max_transform_dist;
+    });
```

Comment:
Nice refactor!

Reply:
This looks like its actually more computation than what we have now. Right now its :
- 1 full iteration and returns the iterator to the closest point
- from the closest point to the robot to the first point that is outside of the lookahead distance. 

So that's 1 full iteration and a `closest -> lookahead distance` sub-iteration.

---

This new one is:
- 1 full iteration
- 1 search to find the minimum element (more or less the same implementation as @rayman's `min_by` except since it wasn't combined with the euclidean, it's an extra bit of CPU O(N-1).
- Then something to turn `min_euclidean_dist` -> `transformation_begin` which I don't understand, https://en.cppreference.com/w/cpp/algorithm/min_element min_element doesn't return the distance, it returns the iterator to the item of the lowest element. So that seems... unnecessary
- another find_if

I think what's there right now is faster, since the second run through of euclidean distances is a pretty short sub-set of the path just from the robot pose to the carrot distance.